exporters/metrics/util/src/main/java/io/opencensus/exporter/metrics/util/QueueMetricProducer.java

```diff
@@ -0,0 +1,196 @@
+  private static final Sampler probabilitySampler = Samplers.probabilitySampler(0.0001);
+
+  private static final String DEFAULT_SPAN_NAME = "PushMetrics";
+  private static final int DEFAULT_BUFFER_SIZE = 32;
+  private static final String ATTRIBUTE_KEY_DROPPED_METRICS = "DroppedMetrics";
+
+  @GuardedBy("monitor")
+  private final Queue<Metric> bufferedMetrics = new LinkedList<>();
+
+  private final int bufferSize;
+  private final String spanName;
+
+  private QueueMetricProducer(int bufferSize, String spanName) {
+    this.bufferSize = bufferSize;
+    this.spanName = spanName;
+  }
+
+  /**
+   * Creates a new {@link QueueMetricProducer}.
+   *
+   * @param options the options for {@link QueueMetricProducer}.
+   * @return a {@code QueueMetricProducer}.
+   * @since 0.20
+   */
+  public static QueueMetricProducer create(Options options) {
+    checkNotNull(options, "options");
+    checkArgument(options.getBufferSize() > 0, "buffer size should be positive.");
+    return new QueueMetricProducer(options.getBufferSize(), options.getSpanName());
+  }
+
+  /**
+   * Pushes {@link Metric}s to this {@link QueueMetricProducer}.
+   *
+   * <p>When buffer of this {@link QueueMetricProducer} is full, the oldest {@link Metric}s will be
+   * dropped.
+   *
+   * @param metrics {@code Metrics} to be added to this {@code QueueMetricProducer}.
+   * @since 0.20
+   */
+  public void pushMetrics(Collection<Metric> metrics) {
+    synchronized (monitor) {
+      Span span =
+          tracer
+              .spanBuilder(spanName)
+              .setRecordEvents(true)
+              .setSampler(probabilitySampler)
+              .startSpan();
+      Scope scope = tracer.withSpan(span);
+      try {
+        bufferedMetrics.addAll(metrics);
+        if (bufferedMetrics.size() > bufferSize) {
+          AttributeValue attributeValue =
+              AttributeValue.longAttributeValue(bufferedMetrics.size() - bufferSize);
+          span.putAttribute(ATTRIBUTE_KEY_DROPPED_METRICS, attributeValue);
+          while (bufferedMetrics.size() > bufferSize) {
+            @SuppressWarnings("Unused")
+            Metric metric = bufferedMetrics.poll();
+          }
+        }
+      } finally {
+        scope.close();
+        span.end();
+      }
+    }
+  }
+
+  @Override
+  public Collection<Metric> getMetrics() {
+    Queue<Metric> metricsToExport;
+    synchronized (monitor) {
+      metricsToExport = new LinkedList<>(bufferedMetrics);
+      bufferedMetrics.clear();
+    }
+    return metricsToExport;
+  }
+
+  /**
+   * Options for {@link QueueMetricProducer}.
+   *
+   * @since 0.20
+   */
+  @AutoValue
+  @Immutable
+  public abstract static class Options {
+
+    Options() {}
+
+    /**
+     * Returns the span name for the {@link Span} created when metrics are pushed.
+     *
+     * @return the span name for the {@link Span} created when metrics are pushed.
+     * @since 0.20
+     */
+    public abstract String getSpanName();
```

Comment:
Don't need this, because metrics will be read via a MetricsReader which does that.